### PR TITLE
V3 Engine Flex

### PR DESCRIPTION
**Pull request by lyaunzbe**
## Description of changes

The V3 Engine Flex contract & interface adds support for the Flex data storage/reference model to the newly updated V3 Engine Core. Namely it also adds "on-chain" as a flex external asset dependency type, allowing users to store string data as bytecode directly on the chain, in addition to the existing options of IPFS or ARWEAVE cid reference.

The base ExternalAssetDependency struct has been modified to keep track of bytecodeAddresses, depending on the dependencyType of the asset (defaults to the zero address for non "on-chain" assets).
The project external asset dependency getter has been  modified to return string data saved as bytecode, if it exists.

The new V3 Engine Flex contract has been added to existing V3 related tests and new, flex specific integration tests have been added that test functionality across ALL depedencyTypes, including the new one this version introduces.

Also, this PR includes V3 Engine changes primarily aimed at reduction of contract size, which where then also applied to this new V3 Engine Flex contract as well. Hence, this PR bumps the latest V3 Engine version to 3.1.2

Engine V3 3.1.1 -> Engine V3 3.1.2 DIFF: https://www.diffchecker.com/r6gIaNJX/
Engine V3 3.1.2 -> Engine Flex V3 3.1.2 DIFF: https://www.diffchecker.com/9iWo55R4/

### CODE REVIEW

**Comment by ryley-o** [comment]
Cc @lyaunzbe - just wanted to note that #449 is intended to work automatically with engine flex as well as engine - I think

**Comment by ryley-o** [comment]
(sorry, didn't mean to close the PR - reopened 😅 😰 )

**Comment by lyaunzbe** [comment]
Mostly ready for review (tests may still be failing due to gas tests needing to be updated in minter tests)

File: contracts/GenArt721CoreV3_Engine_Flex.sol [hunk 1]
```diff
@@ -0,0 +1,2278 @@
...
+     * @notice Returns external asset dependency for project `_projectId` at index `_index`.
+     * If the dependencyType is ONCHAIN, the `data` field will contain the extrated bytecode data and `cid`
+     * will be an empty string. Conversly, for any other dependencyType, the `data` field will be an empty string
+     * and the `bytecodeAddress` will point to the zero address.
+     */
+    function projectExternalAssetDependencyByIndex(
+        uint256 _projectId,
+        uint256 _index
+    ) public view returns (ExternalAssetDependencyWithData memory) {
+        ExternalAssetDependencyType _dependencyType = projects[_projectId]
```
Comment (ryley-o) [inline contracts/GenArt721CoreV3_Engine_Flex.sol:461]: nit: can we eliminate some redundancy (and some redundant hashing and follow pattern like:
```
// first load pointer to external asset dependency in storage
ExternalAssetDependency storage dependency = projects[_projectId].externalAssetDependencies[_index];
// then lookup properties
address bytecodeAddress = dependency.bytecodeAddress
// ... etc.
```

File: contracts/GenArt721CoreV3_Engine_Flex.sol [hunk 2]
```diff
@@ -0,0 +1,2278 @@
...
+     * @param _cidOrData Asset cid (Content identifier) or data string to be translated into bytecode.
+     * @param _dependencyType Asset dependency type.
+     *  0 - IPFS
+     *  1 - ARWEAVE
+     *  2 - ONCHAIN
+     */
+    function updateProjectExternalAssetDependency(
+        uint256 _projectId,
+        uint256 _index,
+        string memory _cidOrData,
```
Comment (ryley-o) [inline contracts/GenArt721CoreV3_Engine_Flex.sol:501]: thinking this should be a generic bytes, with an update our bytecode storage library to handle bytes... let's chat sync about this in some more detail though!

File: contracts/GenArt721CoreV3_Engine_Flex.sol [hunk 3]
```diff
@@ -0,0 +1,2278 @@
...
+        ExternalAssetDependencyType _dependencyType
+    ) external {
+        _onlyUnlockedProjectExternalAssetDependencies(_projectId);
+        _onlyArtistOrAdminACL(
+            _projectId,
+            this.updateProjectExternalAssetDependency.selector
+        );
+        uint24 assetCount = projects[_projectId].externalAssetDependencyCount;
+        require(_index < assetCount, "Asset index out of range");
+        ExternalAssetDependencyType oldDependencyType = projects[_projectId]
```
Comment (ryley-o) [inline contracts/GenArt721CoreV3_Engine_Flex.sol:511]: nit: similar to other comment, I think using the following storage pointer pattern will help with code clarity and gas:
```
ExternalAssetDependency storage dependency = projects[_projectId].externalAssetDependencies[_index];
```

File: contracts/GenArt721CoreV3_Engine_Flex.sol [hunk 4]
```diff
@@ -0,0 +1,2278 @@
...
+                .dependencyType == ExternalAssetDependencyType.ONCHAIN
+        ) {
+            // copy last element bytecode address to index of element to be removed
+            // if we are dealing with an onchain external asset dependency
+            projects[_projectId]
+                .externalAssetDependencies[_index]
+                .bytecodeAddress = projects[_projectId]
+                .externalAssetDependencies[lastElementIndex]
+                .bytecodeAddress;
+        }
```
Comment (ryley-o) [inline contracts/GenArt721CoreV3_Engine_Flex.sol:582]: I believe this operation should already should have been done by the previous statement?

File: contracts/interfaces/0.8.x/IGenArt721CoreContractV3_Engine_Flex.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    // preferredIPFSGateway is a url string
+    function preferredIPFSGateway() external view returns (string memory);
+
+    // preferredArweaveGateway is a url string
+    function preferredArweaveGateway() external view returns (string memory);
+
+    event ExternalAssetDependencyUpdated(
+        uint256 indexed _projectId,
+        uint256 indexed _index,
+        string _cid,
```
Comment (ryley-o) [inline contracts/interfaces/0.8.x/IGenArt721CoreContractV3_Engine_Flex.sol:21]: in natspec comments, can we explicitly call out that this is the empty string if _cid is on-chain data?

File: contracts/GenArt721CoreV3_Engine_Flex.sol [hunk 6]
```diff
@@ -0,0 +1,2278 @@
...
+        _onlyArtistOrAdminACL(
+            _projectId,
+            this.updateProjectExternalAssetDependency.selector
+        );
+        uint24 assetCount = projects[_projectId].externalAssetDependencyCount;
+        address _bytecodeAddress = address(0);
+        // if the incoming dependency type is onchain, we need to write the data to bytecode
+        if (_dependencyType == ExternalAssetDependencyType.ONCHAIN) {
+            _bytecodeAddress = _cidOrData.writeToBytecode();
+            _cidOrData = "";
```
Comment (ryley-o) [inline contracts/GenArt721CoreV3_Engine_Flex.sol:615]: can we add dev comment above this like in the update function?
```
// we don't want to emit data, so we emit the cid as an empty string
```

**Review by ryley-o — APPROVED** [review]
thank you ser 🙏  mostly nits, other than some Q's about potentially using `bytes` instead of `string` for on-chain assets. Let's chat about that detail synchronously if possible 👍

File: contracts/GenArt721CoreV3_Engine_Flex.sol (hunk 2 above)
Comment (ryley-o) [reply contracts/GenArt721CoreV3_Engine_Flex.sol:501]: per discussion, this seems good as-is 👍 happy to discuss in more detail if we learn anything new

**Comment by jakerockland** [comment]
Happy to be second reviewer here @lyaunzbe if that extra set of eyes is helpful given that this PR is a big one, please do feel free to add me as tagged for review once updating based on the feedback from @ryley-o above.

**Review by jakerockland — APPROVED** [review]
This LGTM! Very much appreciate all the meaningful back and forth here in the review and iteration feedback process @lyaunzbe @ryley-o  

So incredibly excited about additional of **on-chain dependencies** as part of the new V3 spec here. 

I think this is a big win for potential partners who want to do more explorative stuff with small data payloads without managing the complexity of an external service like IPFS or Arweave.

Really appreciate being able to pull this into the V3 spec and am excited to be able to start offering this to partners with V3. Big win!! 

cc @snowfro @cooper-temple @ABDruid for visibility about this coming down the pipeline with V3 Flex.

One last ask before merging this @lyaunzbe would just be to update the changelog to clarify that this new functionality of on-chain dependencies is now available in Engine V3 Flex and beyond.

Let's also follow up making sure that this is integrated well with the work done in #485 and #484, perhaps that fits well with our polish week things going on this week.

**Comment by ryley-o** [comment]
>Let's also follow up making sure that this is integrated well with the work done in https://github.com/ArtBlocks/artblocks-contracts/pull/485 and https://github.com/ArtBlocks/artblocks-contracts/pull/484, perhaps that fits well with our polish week things going on this week.

Added an Asana task for that here: https://app.asana.com/0/1203634050572702/1204024755133705/f

**Comment by jakerockland** [comment]
> > Let's also follow up making sure that this is integrated well with the work done in #485 and #484, perhaps that fits well with our polish week things going on this week.
> 
> Added an Asana task for that here: https://app.asana.com/0/1203634050572702/1204024755133705/f

Awesome, thank you Ryley! Merging.

**Pull request merged**